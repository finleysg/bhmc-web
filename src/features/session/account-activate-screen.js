import "./session.scss"

import React from "react"

import { Button } from "components/buttons"
import { ErrorDisplay } from "components/errors"
import { Spinner } from "components/spinners"
import { useAuth } from "context/auth-context"
import { MdAccountCircle } from "react-icons/md"
import { useNavigate, useParams } from "react-router-dom"
import { useAsync } from "utils/use-async"

function AccountActivateScreen() {
  const { uid, token } = useParams()
  const { isError, isLoading, error, run } = useAsync()
  const { activate } = useAuth()
  const navigate = useNavigate()

  React.useEffect(() => {
    run(activate({ uid, token })).catch(() => {})
  }, [uid, token, run, activate])

  const handleLogin = () => {
    navigate("/session/login")
  }

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-green">
          <i>
            <MdAccountCircle />
          </i>
          Account Activation
        </div>
        {isLoading && (
          <div className="login__body">
            <h3 className="text-green">
              <i title="loading">
                <Spinner style={{ marginTop: ".6rem" }} />
              </i>
            </h3>
            <p>Activating your account...</p>
          </div>
        )}
        {!isLoading && isError && (
          <div className="login__body">
            <h3>Activation Failed</h3>
            <ErrorDisplay isError={isError} error={error} />
          </div>
        )}
        {!isLoading && !isError && (
          <div className="login__body">
            <h3>Your Account is Active</h3>
            <p>
              Thank you! Log in now to sign up for an event or update your profile. If you don't
              have a GHIN, you will receive one when you join us for the current golf season.
            </p>
            <Button onClick={handleLogin}>Login</Button>
          </div>
        )}
      </div>
    </div>
  )
}

export { AccountActivateScreen }
