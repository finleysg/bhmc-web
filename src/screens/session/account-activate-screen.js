import "./session.scss"

import React from "react"

import { Button } from "components/buttons"
import { ErrorDisplay } from "components/errors"
import { Spinner } from "components/spinners"
import { useAuth } from "context/auth-context"
import { MdAccountCircle } from "react-icons/md"
import { useNavigate, useParams } from "react-router-dom"
import { useAsync } from "utils/use-async"

function AccountActivateScreen() {
  const { uid, token } = useParams()
  const { isError, isLoading, error, run } = useAsync()
  const { activate } = useAuth()
  const navigate = useNavigate()

  React.useEffect(() => {
    run(activate({ uid, token })).catch(() => {})
  }, [uid, token, run, activate])

  const handleLogin = () => {
    navigate("/session/login")
  }

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-green">
          <i>
            <MdAccountCircle />
          </i>
          Account Activation
        </div>
        {isLoading && (
          <div className="login__body">
            <h3 className="text-green">
              <i title="loading">
                <Spinner style={{ marginTop: ".6rem" }} />
              </i>
            </h3>
            <p>Activating your account...</p>
          </div>
        )}
        {!isLoading && isError && (
          <div className="login__body">
            <h3>Activation Failed</h3>
            <ErrorDisplay error={error} />
          </div>
        )}
        {!isLoading && !isError && (
          <div className="login__body">
            <h3>Your Account is Active</h3>
            <p>Log in now to register for the 2021 club season, or register for an open event.</p>
            <Button onClick={handleLogin}>Login</Button>
          </div>
        )}
      </div>
    </div>
  )
}

export { AccountActivateScreen }
