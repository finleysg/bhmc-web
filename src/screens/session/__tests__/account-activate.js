import { act, render, screen, waitFor } from "@testing-library/react"

import React from "react"

import { rest } from "msw"
import { setupServer } from "msw/node"
import { handlers } from "test/auth-handlers"
import { AuthWrapper, deferred } from "test/test-utils"

import { AccountActivateScreen } from "../account-activate-screen"

const serverURL = process.env.REACT_APP_SERVER_URL
const server = setupServer(...handlers)

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({ uid: "ABC", token: "123" }),
}))

beforeAll(() => {
  server.listen()
})
afterAll(() => server.close())
afterEach(() => {
  server.resetHandlers()
  jest.clearAllMocks()
})

test("successful activation", async () => {
  const { promise, resolve } = deferred()

  render(
    <AuthWrapper>
      <AccountActivateScreen />
    </AuthWrapper>,
  )

  expect(screen.getByLabelText(/loading/i)).toBeInTheDocument()

  await act(() => {
    resolve()
    return promise
  })

  await waitFor(() => expect(screen.queryByText(/your account is active/i)).toBeInTheDocument())
})

test("failed activation", async () => {
  const { promise, resolve } = deferred()
  server.use(
    rest.post(`${serverURL}/auth/users/activation/`, async (req, res, ctx) => {
      const err = ctx.json(["token expired"])
      return res(ctx.status(400), err)
    }),
  )

  render(
    <AuthWrapper>
      <AccountActivateScreen />
    </AuthWrapper>,
  )

  expect(screen.getByLabelText(/loading/i)).toBeInTheDocument()

  await act(() => {
    resolve()
    return promise
  })

  const alert = await screen.findByRole("alert")

  expect(alert.textContent).toMatchSnapshot()
  expect(screen.getByText(/activation failed/i)).toBeInTheDocument()
})
