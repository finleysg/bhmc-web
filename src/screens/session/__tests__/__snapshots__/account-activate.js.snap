// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failed activation 1`] = `"There was an error: [\\"token expired\\"]"`;
